### DISCUSSION TITLE
Port is not open but its specified as exposed port

### DISCUSSION THREAD

**Discussion by patrickbussmann** [community/community#26476] [Actions]
<p>Hello,</p>
<p>I’m a beginner of using GitHub Actions but I’m using docker many years now.</p>
<p>My problem is that the syntax is similar but the behaviors are different.</p>
<p>So lets compare.</p>
<p>My docker-compose is working fine.</p>
<p>I can connect to the 4444 port as perfect as possible.</p>
<p>But the GitHub Action is not able to connect.</p>
<pre><code># docker-compose.yml

version: '3.3'

services:
  selenium-hub:
    image: selenium/hub:3.141.59-20200409
    container_name: selenium-hub
    ports:
      - "4444:4444"

  chrome:
    image: selenium/node-chrome:3.141.59-20200409
    depends_on:
      - selenium-hub
    environment:
      - HUB_HOST=selenium-hub
      - HUB_PORT=4444
</code></pre>
<p>So and now lets check the GitHub Action.</p>
<pre><code># .github/workflows/ci.yml

name: CI

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

jobs:
  build:
    runs-on: ubuntu-latest

    container:
      image: node:12
      options: --network-alias testHost

    services:
      selenium-hub:
        image: selenium/hub:3.141.59
        ports:
          - 4444:4444

      chrome:
        image: selenium/node-chrome:3.141.59
        env:
          HUB_HOST: selenium-hub
          HUB_PORT: 4444

    env:
      SELENIUM_HUB_HOST: selenium-hub
      TEST_HOST: testHost

    steps:
    - uses: actions/checkout@v2

    - name: Install NPM Packages
      run: npm install

    - name: Verify Connection
      env:
        PORT: ${{ job.services.selenium-hub.ports[4444] }}
      run: |
        ping -c 4 testHost
        ping -c 4 selenium-hub
        ping -c 4 localhost

    - name: Run test
      run: npm run test
</code></pre>
<p>But when I run this action workflow now, I’m getting:</p>
<pre><code>PING testHost (172.18.0.2) 56(84) bytes of data.
64 bytes from 87ab046ac468 (172.18.0.2): icmp_seq=1 ttl=64 time=0.029 ms

PING selenium-hub (172.18.0.3) 56(84) bytes of data.
64 bytes from ca5f409a647d4a079dd1fac1b0851069_seleniumhub314159_150933.github_network_ebc2afd0a9664bccb4793a5dc86f64ce (172.18.0.3): icmp_seq=1 ttl=64 time=0.070 ms

PING localhost (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=0.021 ms

UnhandledPromiseRejectionWarning: Error: ECONNREFUSED connect ECONNREFUSED 172.18.0.2:4444
</code></pre>
<p>I searched a lot and also tried to use “sleep 10” seconds before run the “npm run test” command.</p>
<p>But even then its not working.</p>
<p>On my local computer I can use “docker-compose up” and its there and I can connect via localhost:4444 easily.</p>
<p>Whats wrong here in the setup?</p>
<p>Thanks in Advance!</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/patrickbussmann">@patrickbussmann</a> ,</p>
<p>Which step is the connect error occurs on, the " <strong>Verify Connection</strong>" or the " <strong>Run test</strong>"?<br>
I tested with the same steps as you shared, the " <strong>Verify Connection</strong>" step can work fine without any error on my side.<br>
If the error occurs on the " <strong>Run test</strong>" step, please share your repository with us so that we can check more detailed configurations to analyze the root cause.</p>
<p>In addition, I noticed that you have configured the job run in a Docker container ( <strong>image: node:12</strong> ), that means all the steps in this job will run in the container, no steps run directly on the runner machine.</p>
<p>What you need to pay attention to (more details, see docs about <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/about-service-containers" rel="nofollow noopener"><strong>service containers</strong></a>):<br>
<strong>1)</strong> Communication between a job and its service containers is different depending on whether a job runs directly on the runner machine or in a container.</p>
<p><strong>2)</strong> When you run jobs in a container, GitHub connects service containers to the job using Docker’s user-defined bridge networks. Running the job and services in a container simplifies network access. You can access a service container using the label you configure in the workflow. You don’t need to configure any ports for service containers. By default, all containers that are part of the same Docker network expose all ports to each other, and no ports are exposed outside of the Docker network.</p>
<p><strong>3)</strong> When running jobs directly on the runner machine, you can access service containers using <strong>localhost:&lt;port&gt;</strong> or <strong>127.0.0.1:&lt;port&gt;</strong>. GitHub configures the container network to enable communication from the service container to the Docker host.</p>

**Comment by patrickbussmann** [comment]
<p>Yes great, thank you.</p>
<p>In the night I had the same idea.</p>
<p>I searched again for the problem and found out that the documentation says there is an difference between “on-host runner” and “containers”.</p>
<p>And I tought the whole time I’m using it on the host - until I saw the container option with node.</p>
<p>So I deleted it and added an install NPM step so that I can directly access 4444 port.</p>
<p>And … et voila, it works fine.</p>
<p>But then I wake up and saw that the build was running 6 hours.</p>
<p>Thats a bit bad, so I decreased the timeout now to 5 minutes.</p>
<p>Thank you for helping.</p>

**Comment by danielfrg** [comment]
<p>Hi,</p>
<p>Could you share your updated action.yml file?</p>
<p>I am trying this one but its failing in the verify step:</p>
<pre><code>name: CI

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

jobs:
  build:
    runs-on: ubuntu-latest

    services:
      selenium-hub:
        image: selenium/hub:3.141.59
        ports:
          - 4444:4444

      chrome:
        image: selenium/node-chrome:3.141.59
        env:
          HUB_HOST: selenium-hub
          HUB_PORT: 4444

    env:
      SELENIUM_HUB_HOST: selenium-hub
      TEST_HOST: testHost

    steps:
    - uses: actions/checkout@v2

    - name: Verify Connection
      env:
        PORT: ${{ job.services.selenium-hub.ports[4444] }}
      run: |
        ping -c 4 localhost
        ping -c 4 selenium-hub
</code></pre>
<p>Error on verify connection step:</p>
<pre><code>PING localhost (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=0.032 ms
64 bytes from localhost (127.0.0.1): icmp_seq=2 ttl=64 time=0.038 ms
64 bytes from localhost (127.0.0.1): icmp_seq=3 ttl=64 time=0.041 ms
64 bytes from localhost (127.0.0.1): icmp_seq=4 ttl=64 time=0.037 ms

--- localhost ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3036ms
rtt min/avg/max/mdev = 0.032/0.037/0.041/0.003 ms
ping: selenium-hub: Temporary failure in name resolution
##[error]Process completed with exit code 2.
</code></pre>
<p>Link to the job if its useful: <a href="https://github.com/danielfrg/jupyter-flex/runs/614305160?check_suite_focus=true" rel="nofollow noopener">https://github.com/danielfrg/jupyter-flex/runs/614305160?check_suite_focus=true</a></p>
<p>Thanks!</p>

**Comment by danielfrg** [comment]
<p>Nevermind! I missed the part where he explained that when running on the host you can connect to the services using <code>localhost:port</code>.</p>
<p>Everything is working now :D </p>
